### PR TITLE
Fix Rule-handling in PEtab import

### LINKED ISSUE #1750: AMICI does not handle conditions for parameters governed by rate rules
Hi all,
trying to load the attached PETab problem

[icg_petab_v2.zip](https://github.com/AMICI-dev/AMICI/files/8381643/icg_petab_v2.zip)

but getting the following error.
```
amici.sbml_import.SBMLException: Cannot turn parameter IVDOSE_icg into a constant/fixed parameter since it either has an initial assignment or is the target of an assignment or rate rule.
```

This is a simple model which has the following rate rule
```
dIVDOSE_idg/dt = -iv_icg*Mr_icg + Ri_icg
```
As part of my conditions I want to set `IVDOSE_icg[t=0] = value`. I assume you handle parameters different then species and compartments, but this is not correct. You have to split the set of parameters in constant parameters and parameters governed by rate rules. Parameters governed by rate rules are just state variables in the ODE and everything which is allowed for species state variables and compartment state variables must also work for parameter state variables!

I.e. the more correct way for the condition table description in 
https://petab.readthedocs.io/en/latest/documentation_data_format.html#condition-table
should be

>The condition table specifies constant parameters, or initial values of species, compartments or parameters for specific simulation conditions (generally corresponding to different experimental conditions).

instead of
>The condition table specifies parameters, or initial values of species and compartments for specific simulation conditions (generally corresponding to different experimental conditions).

Could you please support the setting of initial conditions for parameters governed by rate rules? This mathematical construct occurs in many models and it would be completely unnatural to encode such parameters as `dummy` species or compartments just to get PETab to work.

Code example with full error stack below.
```
# Import yaml
petab_problem = petab.Problem.from_yaml(yaml_file)

# compile the petab problem to an AMICI ODE model
amici_model = amici.petab_import.import_petab_problem(petab_problem)
```
results in

```
/home/mkoenig/.virtualenvs/pkdb_models/bin/python /home/mkoenig/git/pkdb_models/pkdb_models/models/icg/fit_petab/icg_amici.py
2022-03-30 16:59:37.272 - amici.petab_import - INFO - Importing model ...
2022-03-30 16:59:37.273 - amici.petab_import - INFO - Model name is 'icg_body'.
Writing model code to '/home/mkoenig/git/pkdb_models/pkdb_models/models/icg/fit_petab/amici_models/icg_body'.
2022-03-30 16:59:37.274 - amici.petab_import - INFO - Species: 12
2022-03-30 16:59:37.274 - amici.petab_import - INFO - Global parameters: 75
2022-03-30 16:59:37.274 - amici.petab_import - INFO - Reactions: 16
2022-03-30 16:59:37.329 - amici.petab_import - INFO - Observables: 5
2022-03-30 16:59:37.329 - amici.petab_import - INFO - Sigmas: 5
2022-03-30 16:59:37.333 - amici.petab_import - DEBUG - Adding output parameters to model: OrderedDict([('noiseParameter1_Cve_icg', None), ('noiseParameter1_Chv_icg', None), ('noiseParameter1_Car_icg', None), ('noiseParameter1_ER_icg', None), ('noiseParameter1_LI__ICGLI2BI', None)])
2022-03-30 16:59:37.333 - amici.petab_import - DEBUG - Adding initial assignments for []
2022-03-30 16:59:37.333 - amici.petab_import - DEBUG - Condition table: (16, 5)
2022-03-30 16:59:37.333 - amici.petab_import - DEBUG - Fixed parameters are ['BW', 'Ri_icg', 'IVDOSE_icg']
2022-03-30 16:59:37.333 - amici.petab_import - INFO - Overall fixed parameters: 3
2022-03-30 16:59:37.333 - amici.petab_import - INFO - Variable parameters: 77
2022-03-30 16:59:37.390 - amici.sbml_import - INFO - Finished gathering local SBML symbols       ++ (5.58E-02s)
Traceback (most recent call last):
  File "/home/mkoenig/git/pkdb_models/pkdb_models/models/icg/fit_petab/icg_amici.py", line 28, in <module>
    amici_model = amici.petab_import.import_petab_problem(petab_problem)
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/petab_import.py", line 305, in import_petab_problem
    import_model_sbml(
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/logging.py", line 194, in wrapper_timer
    rval = func(*args, **kwargs)
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/petab_import.py", line 579, in import_model_sbml
    sbml_importer.sbml2amici(
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/sbml_import.py", line 358, in sbml2amici
    self._process_sbml(constant_parameters)
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/logging.py", line 194, in wrapper_timer
    rval = func(*args, **kwargs)
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/sbml_import.py", line 408, in _process_sbml
    self._process_parameters(constant_parameters)
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/logging.py", line 194, in wrapper_timer
    rval = func(*args, **kwargs)
  File "/home/mkoenig/.virtualenvs/pkdb_models/lib/python3.9/site-packages/amici/sbml_import.py", line 845, in _process_parameters
    raise SBMLException(
amici.sbml_import.SBMLException: Cannot turn parameter IVDOSE_icg into a constant/fixed parameter since it either has an initial assignment or is the target of an assignment or rate rule.

Process finished with exit code 1
```

### PR DESCRIPTION
**Pull request by dweindl**
RateRules-targets in the condition table were previously ignored.

Closes #1750

Also fixes a bug where InitialAssignments were not honored for elements that occurred as column in the condition table and were set to `NaN` (i.e. "use model value").

### CODE REVIEW

**Comment by matthiaskoenig** [comment]
@dweindl Thanks for the work on the issue.

As I understand this fixes the rate rule issue. Would it be possible to merge this in develop?
Best M

File: .github/workflows/test_petab_test_suite.yml [hunk 1]
```diff
@@ -63,9 +63,9 @@ jobs:
       # retrieve test models
       - name: Download and install PEtab test suite
         run: |
-          git clone --depth 1  --branch develop https://github.com/PEtab-dev/petab_test_suite \
+          git clone --depth 1 --branch test_raterules https://github.com/PEtab-dev/petab_test_suite \
             && source ./build/venv/bin/activate \
-            && cd petab_test_suite && pip3 install -e . && cd ..
+            && cd petab_test_suite && pip3 install -e . && pip install git+https://github.com/PEtab-dev/libpetab-python@develop
```
Comment (dweindl) [inline .github/workflows/test_petab_test_suite.yml:68]: TODO

File: python/amici/petab_import.py [hunk 2]
```diff
@@ -514,12 +515,11 @@ def import_model_sbml(
...
     # <BeginWorkAround>
+
     initial_states = [col for col in condition_df
-                      if sbml_model.getSpecies(col) is not None]
-    initial_sizes = [col for col in condition_df
-                     if sbml_model.getCompartment(col) is not None]
+                      if element_is_state(sbml_model, col)]
     fixed_parameters = []
-    if len(initial_states) or len(initial_sizes):
+    if len(initial_states):
```
Comment (FFroehlich) [inline python/amici/petab_import.py:520]:
```suggestion
    if initial_states:
```

File: python/amici/petab_import.py [hunk 3]
```diff
@@ -536,8 +536,8 @@ def import_model_sbml(
         logger.debug("Adding preequilibration indicator "
                      f"constant {PREEQ_INDICATOR_ID}")
     logger.debug("Adding initial assignments for "
-                 f"{initial_sizes + initial_states}")
-    for assignee_id in chain(initial_sizes, initial_states):
+                 f"{initial_states}")
+    for assignee_id in (initial_states):
```
Comment (FFroehlich) [inline python/amici/petab_import.py:538]:
```suggestion
    for assignee_id in initial_states:
```

File: python/amici/petab_objective.py [hunk 4]
```diff
@@ -352,14 +352,24 @@ def create_parameter_mapping_for_condition(
...
+        def _set_initial_state(condition_id, element_id, init_par_id,
+                               par_map, scale_map):
             value = petab.to_float_if_float(
-                petab_problem.condition_df.loc[condition_id, species_id])
+                petab_problem.condition_df.loc[condition_id, element_id])
             if pd.isna(value):
-                value = get_species_initial(
-                    petab_problem.sbml_model.getSpecies(species_id)
-                )
+                element = petab_problem.sbml_model.getElementBySId(element_id)
```
Comment (FFroehlich) [inline python/amici/petab_objective.py:360]: hmm this doesn't account for InitialAssignments right? (was already the case for species before, but I am not sure thats correct in the first place)

**Review by FFroehlich — APPROVED** [review]
👍

File: python/amici/petab_import.py [hunk 5]
```diff
@@ -86,7 +86,8 @@ def get_fixed_parameters(
         # remove overridden parameters (`object`-type columns)
         fixed_parameters = [p for p in fixed_parameters
                             if condition_df[p].dtype != 'O'
-                            and sbml_model.getParameter(p) is not None]
+                            and sbml_model.getParameter(p) is not None
+                            and sbml_model.getRuleByVariable(p) is None]
```
Comment (FFroehlich) [inline python/amici/petab_import.py:90]: uuuuh, so what happens with rules here that don't have a RateRule qualifier?

Comment (dweindl) [reply python/amici/petab_import.py:90]: This is about whether these parameters should be considered as AMICI-constant-parameters. Any parameters that are rule targets cannot be constants. Or am I missing something?

File: python/amici/petab_objective.py (hunk 4 above)
Comment (dweindl) [reply python/amici/petab_objective.py:360]: Yes, you are right. Good catch. Should add a test case.

Comment (dweindl) [reply python/amici/petab_objective.py:360]: Fixed

**Review by FFroehlich — APPROVED** [review]
👍

**Comment by dweindl** [comment]
> Would it be possible to merge this in develop?

@matthiaskoenig : now available on pypi - `amici>=0.11.28`

**Comment by matthiaskoenig** [comment]
@dweindl Thanks. This part works, just moving on to the next issue with model import ;) See #1779.

**Pull request merged**